service_container/service_decoration.rst

```diff
@@ -313,4 +313,117 @@ The generated code will be the following::
+
+            <services>
+                <service id="Foo"/>
+
+                <service id="Bar" decorates="Foo" decoration-on-invalid="ignore">
+                    <argument type="service" id="Bar.inner"/>
+                </service>
+            </services>
+        </container>
+
+    .. code-block:: php
+
+        // config/services.php
+        namespace Symfony\Component\DependencyInjection\Loader\Configurator;
+
+        use Symfony\Component\DependencyInjection\ContainerInterface;
+
+        return function(ContainerConfigurator $configurator) {
+            $services = $configurator->services();
+
+            $services->set(Foo::class);
+
+            $services->set(Bar::class)
+                ->decorate(Foo::class, null, 0, ContainerInterface::IGNORE_ON_INVALID_REFERENCE)
+                ->args([ref(Bar::class.'.inner')]);
+
+            $services->set(Baz::class)
+                ->private()
+                ->decorate(Foo::class, null, 1)
+                ->args([ref(Baz::class.'.inner')]);
+        };
+
+Three different behaviors are available:
+
+* ``exception``: When used, a ``ServiceNotFoundException``
+  will be thrown telling that decorator's dependency is missing. (default)
+* ``ignore``: Using that behavior, the container will remove either decorator
+  and decorated services.
+* ``null``: When used, the container will keep the decorator
+  service and will set the decorated one to ``null``.
+
+.. tip::
+
+    When using ``null``, you may update the decorator's constructor in
+    order to make decorated dependency nullable. In this way, you will be able to
+    check the presence of the decorated service in other methods.
+
+    .. code-block:: yaml
+
+        App\Service\DecoratorService:
+            decorates: Acme\OptionalBundle\Service\OptionalService
+            decoration_on_invalid: null
+            arguments: ['@App\Service\DecoratorService.inner']
+
+    .. code-block: php
+
+        namespace App\Service;
+
+        use Acme\OptionalBundle\Service\OptionalService;
+
+        class DecoratorService
+        {
+            private $decorated;
+
+            public function __construct(?OptionalService $decorated)
+            {
+                $this->decorated = $decorated;
+            }
+
+            public function tellInterestingStuff(): string
+            {
+                if (!$this->decorated) {
+                    return 'Just one interesting thing';
+                }
+
+                return $this->decorated->tellInterestingStuff().' + one more interesting thing';
+            }
+        }
+
+.. tip::
+
+    Sometimes, you may want to add a compiler pass that creates service
+    definitions on the fly. If you want to decorate such a service,
+    be sure that your compiler pass is registered with ``PassConfig::TYPE_BEFORE_OPTIMIZATION``
+    type so that the decoration pass will be able to find the created services.
+
```

Comment:
Missing a `.. versionadded::` directive about the `decoration_on_invalid` option.